lib/lightning/usage_tracking/workflow_metrics_query.ex

```diff
@@ -0,0 +1,31 @@
+defmodule Lightning.UsageTracking.WorkflowMetricsQuery do
+  @moduledoc """
+  Query module to support workflow metrics while allowing for fewer test
+  permutations.
+
+
+  """
+  import Ecto.Query
+
+  alias Lightning.Run
+
+  def workflow_runs(%{id: workflow_id}) do
+    from r in Run,
+      join: w in assoc(r, :workflow),
+      on: w.id == ^workflow_id
+  end
+
+  def runs_finished_on(query, date) do
+    start_of_day = DateTime.new!(date, ~T[00:00:00])
+
+    start_of_next_day =
+      date
+      |> Date.add(1)
+      |> DateTime.new!(~T[00:00:00])
+
+    from r in query,
+      where: not is_nil(r.finished_at),
+      where: r.finished_at >= ^start_of_day,
+      where: r.finished_at < ^start_of_next_day
```

Comment:
@rorymckinley I just realised we don't have an index on `finished_at`, while this wouldn't be a cause of the memory issue, it probably is a performance one right?

Reply:
@stuartc If you view this portion of the query in isolation, then yes it could have an impact. However, the overall reporting query also filters runs by workflow so there is a chance Postgres would use the those indexes when retrieving the records.


Of course, this also depends on the nature of the data in the Db at the time and whether the query planner has a case of the Mondays :D. It is also not to say that Postres would not prefer to use the `finished_at`  if it had an index.

Happy to run an `EXPLAIN ANALYZE` on prod if you like?

Reply:
@stuartc And of course the counterargument is: Do you want to potentially slow down run insertions to allow the reports to complete quicker?

Reply:
Yeah please do, find a project with lots of runs and do a check - we're also looking at adding indexes for some tables that get hammered when retention jobs come along and trigger cascade deletes - so it's been on my mind.

I'll also take a look around for things that also use `finished_at` and co for things like ordering. So I'm on your side here, we don't want to add indexes for the sake of it - which ironically is the same practice that has created some of the performance issues we've observed.

Reply:
I think holding off on indexing until you see performance issues is the accepted approach, I just tend to see them too late :).